File: 20-valid-parentheses/20-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        complementry = {
          "{":"}",
          "(":")",
          "[":"]",
        }
        if len(s)%2 != 0:
            return False
        for char in s:
            if char in complementry.keys():
                stack.append(complementry[char])
            elif not len(stack):
                return False
            elif char != stack.pop():
                return False
        return len(stack) == 0
        

File: 20-valid-parentheses/20-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
    stack<char> st;
    bool balanced=true;
    for(auto c: s){
        if (c == '}' || c == ']' || c == ')'){
            if(!st.empty()) {
                char adj = st.top();
                st.pop();
                if ((c == '}' && adj != '{') || (c == ']' && adj != '[') || (c == ')' && adj != '(')) {
                    return false;
                }
            }else{
                return false;
            }
        }else{
            st.push(c);
        }
    }
    if(balanced && st.empty()) return true;
    else return false;
    return 1; 
    }
};